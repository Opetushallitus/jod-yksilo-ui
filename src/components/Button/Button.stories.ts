import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

const meta = {
  title: 'Primitives/Button',
  component: Button,
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Base: Story = {
  args: {
    label: 'Base Button',
    variant: 'base',
    onClick: () => {
      // Do nothing
    },
  },
};

export const Primary: Story = {
  args: {
    label: 'Primary Button',
    variant: 'primary',
    onClick: () => {
      // Do nothing
    },
  },
};

export const Outlined: Story = {
  args: {
    label: 'Outlined Button',
    variant: 'outlined',
    onClick: () => {
      // Do nothing
    },
  },
};

export const Text: Story = {
  args: {
    label: 'Text Button',
    variant: 'text',
    onClick: () => {
      // Do nothing
    },
  },
};
