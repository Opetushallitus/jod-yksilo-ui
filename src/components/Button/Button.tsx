import classNames from 'classnames';
/**
 * base = default button to be used
 * primary = should be only appear once in a page
 * text = for less-pronounced actions
 */
export type ButtonVariant = 'base' | 'primary' | 'outlined' | 'text';

export interface ButtonProps {
  /* Text shown on the button*/
  label: string;
  /* Callback fired on tap/click of the button */
  onClick: () => void;
  /** 
   Variant of the button 
   */
  variant?: ButtonVariant;
  /* Button disabled for any actions */
  disabled?: boolean;
}

export const Button = ({ label, onClick, variant = 'base', disabled = false }: ButtonProps) => {
  return (
    <button
      disabled={disabled}
      type="button"
      onClick={onClick}
      className={classNames('jod-button', 'm-2 px-4 py-2 font-bold', {
        'bg-jod-base text-jod-white': variant === 'base',
        'bg-jod-primary text-jod-white': variant === 'primary',
        'border-jod-black text-jod-black': variant === 'outlined',
        'text-jod-black': variant === 'text',
        'cursor-not-allowed opacity-50': disabled,
      })}
    >
      {label}
    </button>
  );
};
