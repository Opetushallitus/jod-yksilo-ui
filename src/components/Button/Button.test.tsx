import { describe, expect, it, test, vi } from 'vitest';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from './Button';

describe('Snapshot testing', () => {
  test('Base button', () => {
    const { container } = render(<Button label="Base button" variant="base" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Primary button', () => {
    const { container } = render(<Button label="Primary button" variant="primary" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Outlined button', () => {
    const { container } = render(<Button label="Outlined button" variant="outlined" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('Text button', () => {
    const { container } = render(<Button label="Text button" variant="text" onClick={vi.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('renders the button', () => {
  render(<Button label="Rendered button" onClick={vi.fn()} />);
  expect(true).toBeTruthy();
});

it('has the correct label', () => {
  render(<Button label="Label" onClick={vi.fn()} />);
  expect(screen.getByRole('button', { name: 'Label' })).not.toBeNull();
});

it('calls the callback on click', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<Button label="Test callback" onClick={mockCallback} />);
  const button = screen.getByRole('button', { name: 'Test callback' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

it('does not call the callback while disabled', async () => {
  const mockCallback = vi.fn();
  const user = userEvent.setup();

  render(<Button label="Disabled Button" onClick={mockCallback} disabled />);
  const button = screen.getByRole('button', { name: 'Disabled Button' });
  await user.click(button);
  expect(mockCallback).toHaveBeenCalledTimes(0);
});
