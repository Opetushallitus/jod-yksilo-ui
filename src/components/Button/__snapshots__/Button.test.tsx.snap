// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Base button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-base text-jod-white"
        type="button"
      >
        Base button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="jod-button m-2 px-4 py-2 font-bold bg-jod-base text-jod-white"
      type="button"
    >
      Base button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Snapshot testing > Outlined button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-base text-jod-white"
        type="button"
      >
        Base button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-primary text-jod-white"
        type="button"
      >
        Primary button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold border-jod-black text-jod-black"
        type="button"
      >
        Outlined button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="jod-button m-2 px-4 py-2 font-bold border-jod-black text-jod-black"
      type="button"
    >
      Outlined button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Snapshot testing > Primary button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-base text-jod-white"
        type="button"
      >
        Base button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-primary text-jod-white"
        type="button"
      >
        Primary button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="jod-button m-2 px-4 py-2 font-bold bg-jod-primary text-jod-white"
      type="button"
    >
      Primary button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Snapshot testing > Text button 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-base text-jod-white"
        type="button"
      >
        Base button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold bg-jod-primary text-jod-white"
        type="button"
      >
        Primary button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold border-jod-black text-jod-black"
        type="button"
      >
        Outlined button
      </button>
    </div>
    <div>
      <button
        class="jod-button m-2 px-4 py-2 font-bold text-jod-black"
        type="button"
      >
        Text button
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="jod-button m-2 px-4 py-2 font-bold text-jod-black"
      type="button"
    >
      Text button
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
